src/actions/marketo/marketo.ts

```diff
@@ -43,46 +35,52 @@ export class MarketoAction extends Hub.Action {
   supportedFormats = [Hub.ActionFormat.JsonDetail]
   supportedFormattings = [Hub.ActionFormatting.Unformatted]
   supportedVisualizationFormattings = [Hub.ActionVisualizationFormatting.Noapply]
+  usesStreaming = true
 
   async execute(request: Hub.ActionRequest) {
-    if (!(request.attachment && request.attachment.dataJSON)) {
-      throw "No attached json."
-    }
-
     if (!request.formParams.campaignID) {
       throw "Missing Campaign ID."
     }
 
-    const requestJSON = request.attachment.dataJSON
-    if (!requestJSON.fields || !requestJSON.data) {
-      throw "Request payload is an invalid format."
-    }
-    const fieldMap = this.tagMap(requestJSON)
-    const leadList = this.leadList(fieldMap, requestJSON)
-
-    // Push leads into Marketo and affiliate with a campaign
-    const numLeadsAllowedPerCall = 300
-    const chunked = MarketoAction.chunkify(leadList, numLeadsAllowedPerCall)
+    let fieldset: Hub.Field[] = []
+    const fieldMap: any = {}
+    const errors: Error[] = []
     const marketoClient = this.marketoClientFromRequest(request)
-    const errors: {message: string}[] = []
-    for (const chunk of chunked) {
-      try {
-        const newLeads = await marketoClient.lead.createOrUpdate(chunk)
-        const justIDs = newLeads.result.map((lead: {id: number}) => ({ id: lead.id }))
-        await marketoClient.campaign.request(request.formParams.campaignID, justIDs)
-      } catch (e) {
-        errors.push(e)
-      }
-    }
 
-    let response
+    await request.streamJsonDetail({
+      onFields: (fields) => {
+        fieldset = Hub.allFields(fields)
+        let hasTagMap
+        for (const field of fieldset) {
+          hasTagMap = field.tags.map((tag: string) => {if (tag.startsWith("marketo:")) {
+            return tag.split("marketo:")[1]
+          }})
+          fieldMap[field.name] = hasTagMap.find((tag: string | undefined) => tag !== undefined)
+        }
+      },
+      onRow: async (row) => {
```

Comment:
hmm... I'm not sure we can do `async` here because the streaming code won't await it...

Reply:
Does marketo support a batch API?

Reply:
Yeah. They'll accept up to 300 records in the lead.createOrUpdate.

What is the right way to handle the async process? We need the createOrUpdate result ids prior to adding the ids to the campaign.

Reply:
I'm not sure – using it here will _work_ but since it's not waited on the action will return success before it finishes. We can probably make `streamJsonDetail` optionally allow async callbacks and await on them.

Reply:
(n.b. it's actually 100 records per call, not 300)